x-pack/plugins/entity_manager/server/lib/queries/utils.ts

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import { Entity } from '@kbn/entities-schema';
+import { ESQLSearchResponse } from '@kbn/es-types';
+import { uniq } from 'lodash';
+
+function mergeEntities(entity1: Entity, entity2: Entity): Entity {
+  const merged: Entity = {
+    ...entity1,
+    'entity.last_seen_timestamp': new Date(
+      Math.max(
+        Date.parse(entity1['entity.last_seen_timestamp']),
+        Date.parse(entity2['entity.last_seen_timestamp'])
+      )
+    ).toISOString(),
+  };
+
+  for (const [key, value] of Object.entries(entity2).filter(([_key]) =>
+    _key.startsWith('metadata.')
+  )) {
+    if (merged[key]) {
+      merged[key] = uniq([
+        ...(Array.isArray(merged[key]) ? merged[key] : [merged[key]]),
+        ...(Array.isArray(value) ? value : [value]),
+      ]);
+    } else {
+      merged[key] = value;
+    }
+  }
+  return merged;
+}
+
+export function mergeEntitiesList(entities: Entity[]): Entity[] {
+  const instances: { [key: string]: Entity } = {};
+
+  for (let i = 0; i < entities.length; i++) {
+    const entity = entities[i];
+    const id = entity['entity.id'];
+
+    if (instances[id]) {
+      instances[id] = mergeEntities(instances[id], entity);
+    } else {
+      instances[id] = entity;
+    }
+  }
+
+  return Object.values(instances);
+}
+
+export async function runESQLQuery<T>({
```

Comment:
Can we use the Observability ES client instead?

```
const esClient = createObservabilityEsClient({
    client: clusterClient.asInternalUser,
    logger,
    plugin: '@kbn/entityManager-plugin',
  });
```

It has an `esql` function which also has built in APM instrumentation and handles converting the columnar response to objects.

Reply:
used that before but it was updated to unflatten the objects which makes our merging logic harder. we can maybe update the client to unflatten only when specified but until then let's keep it like this

Reply:
Dario has a PR up right now that reverts that change so you can set which transform mode to use, could we wait for that PR to land?

Reply:
I'll make the change if it's merged before this PR is ready to go